### PR TITLE
[Generate] Remove attention_mask and integrate model_main_input_name

### LINKED ISSUE #13825: Consistent speech model input names for the Seq2SeqTrainer generate function
# 🚀 Feature request

Could we maybe have a consistent naming convention for speech models? So far we have:
- [`input_features`](https://huggingface.co/transformers/model_doc/speech_to_text.html#speech2textforconditionalgeneration)
- [`input_values`](https://huggingface.co/transformers/model_doc/wav2vec2.html#wav2vec2forctc)
- [`input_ids`](https://huggingface.co/transformers/model_doc/speech_to_text_2.html#speech2text2forcausallm)

From what I can tell, these are mostly the same for the purposes of how the `Seq2SeqTrainer` interprets them.

## Motivation

This would prevent the need for custom `Seq2SeqTrainer` classes and would make training more modular.

## Your contribution

A change in param names would do the trick but could break a lot of code. Alternatively adding the capability to accept different key values in the `generate`  function [here](https://github.com/huggingface/transformers/blob/41436d3dfb98e0d17f018db29790b65663358edf/examples/legacy/seq2seq/seq2seq_trainer.py#L219) would work too using a (clunky) mapping such as `INPUT_MAPPING_LABELS = {"input_features": "input_ids", "input_values": "input_ids", "input_ids": "input_ids"}`.

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Final clean-up of generate leveraging the `main_input_name` PR: https://github.com/huggingface/transformers/pull/14803 to:
- Remove `attention_mask` hack for vision models
- clean up the trainer_seq2seq to fix: https://github.com/huggingface/transformers/issues/13825
- clean up hard-coded input names in `generate`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/deit/modeling_deit.py [hunk 1]
```diff
@@ -478,7 +478,6 @@ class PreTrainedModel
     def forward(
         self,
         pixel_values=None,
-        attention_mask=None,
```
Comment (patrickvonplaten) [inline src/transformers/models/deit/modeling_deit.py:481]: model doesn't have `attention_mask`

File: src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.py [hunk 2]
```diff
@@ -176,7 +181,7 @@ class SpeechEncoderDecoderModel(PreTrainedModel):
     """
     config_class = SpeechEncoderDecoderConfig
     base_model_prefix = "speech_encoder_decoder"
-    main_input_name = "input_values"
+    main_input_name = "inputs"
```
Comment (patrickvonplaten) [inline src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.py:184]: SpeechEncoderDecoder is can have multiple different input formats -> change main input name to `inputs`. This has no breaking changes

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -349,9 +350,6 @@ class BeamSampleEncoderDecoderOutput(ModelOutput):
 BeamSampleOutput = Union[BeamSampleEncoderDecoderOutput, BeamSampleDecoderOnlyOutput]
 
 
-ENCODER_MODEL_INPUT_NAMES = ["input_ids", "inputs_embeds", "input_values", "input_features", "pixel_values"]
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:352]: Let's remove hard-coded file-names here

File: src/transformers/generation_utils.py [hunk 4]
```diff
@@ -363,58 +361,69 @@ def _prepare_model_inputs(
...
-            # -> return input
-            model_input_name = None
-            return inputs, model_input_name, model_kwargs
-        elif inputs is not None and len(model_input_kwarg_names) > 0:
-            # 2. `inputs` are passed as well as model-specific keyword inputs
-            # -> not allowed, raise Error
+        # 1. retrieve all kwargs that are non-None or non-model input related.
+        # some encoder-decoder models have different names for model and encoder
+        if (
+            self.config.is_encoder_decoder
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:371]: this is sadly a bit hacky, but there isn't really another way as composite models like `EncoderDecoder`, `RAG`, `SpeechEncoderDecoder`, `VisionEncoderDecoder` can have different kinds of encoders and thus input names

File: src/transformers/generation_utils.py [hunk 5]
```diff
@@ -363,58 +361,69 @@ def _prepare_model_inputs(
...
-                )
-            return input_tensor, model_input_name, model_kwargs
-        else:
-            # 5. no `inputs` are passed and multiple model-specific keyword inputs
-            # -> not allowed, raise Error
+        elif inputs_kwarg is not None:
+            inputs = inputs_kwarg
+
+        # 3. models with `input_ids` can also make use of `inputs_embeds`
+        if self._can_retrieve_inputs_from_name(inputs, "inputs_embeds", model_kwargs):
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:394]: `"inputs_embeds"` is the only input which is not a "main" input name, but nevertheless can be used for `generation` in encoder-decoder models

File: src/transformers/generation_utils.py [hunk 6]
```diff
@@ -1013,12 +1015,13 @@ def generate(
         model_kwargs["output_hidden_states"] = output_hidden_states
         model_kwargs["use_cache"] = use_cache
 
-        if model_kwargs.get("attention_mask", None) is None:
+        has_attention_mask = "attention_mask" in set(inspect.signature(self.forward).parameters.keys())
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:1018]: don't see a better way at the moment of verfying if model has `attention_mask` or not. But don't think calling `inspect` here once is a problem no?

File: src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.py:184]: Just like `perceiver`

File: tests/test_modeling_vision_encoder_decoder.py [hunk 7]
```diff
@@ -64,14 +64,7 @@ def get_pretrained_model_and_inputs(self):
         pass
 
     def check_encoder_decoder_model_from_pretrained_configs(
```
Comment (patrickvonplaten) [inline tests/test_modeling_vision_encoder_decoder.py:66]: vision encoders don't need the "dummy" attention_mask anymore

File: src/transformers/generation_utils.py (hunk 6 above)
Comment (sgugger) [reply src/transformers/generation_utils.py:1018]: Don't see a problem with that.

File: src/transformers/trainer_seq2seq.py [hunk 8]
```diff
@@ -161,11 +161,17 @@ def prediction_step(
...
+        # varying model input names
+        if hasattr(self.model, "encoder") and self.model.encoder.main_input_name != self.model.main_input_name:
+            generation_inputs = inputs[self.model.encoder.main_input_name]
+        else:
+            generation_inputs = inputs[self.model.main_input_name]
 
         generated_tokens = self.model.generate(
-            **generation_inputs,
+            generation_inputs,
+            attention_mask=inputs.get("attention_mask", None),
```
Comment (sgugger) [inline src/transformers/trainer_seq2seq.py:174]: Can you copy that logic to the newly added `KerasMetricCallback` in keras_callbacks? It has a path with `predict_with_generate` that should use the same logic as here. I think we probably can remove the tokenizer argument/attribute of that callback as a result.

**Review by sgugger — APPROVED** [review]
Great addition! We just need to also port this on the TF side.

File: src/transformers/keras_callbacks.py [hunk 9]
```diff
@@ -97,10 +97,11 @@ def __init__(
         self.predict_with_generate = predict_with_generate
         self.output_cols = output_cols
         self.metric_fn_kwargs = metric_fn_kwargs or dict()
-        if tokenizer is not None:
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:100]: That was the only use of the `tokenizer`, so let's remove it from the init (PR was merged yesterday so we can break slightly like this :-) )

**Comment by cgawron** [comment]
The changes to `Seq2SeqTrainer` also fix a problem with `predict_with_generate` for `VisionEncoderDecoder`.

With 4.15.0 the following does not work with `VisionEncoderDecoder`
```Python
def compute_metrics(eval_prediction):
    preds = tokenizer.batch_decode(eval_prediction.predictions, skip_special_tokens=True)
    refs = tokenizer.batch_decode(eval_prediction.label_ids, skip_special_tokens=True)
```

**Comment by patrickvonplaten** [comment]
Hey @cgawron,

Do you have an issue with current master or is everything resolved? :-) Happy to make some more changes to make VisionEncoderDecoder work, but if I understand correctly everything is fine now no?

**Comment by cgawron** [comment]
@patrickvonplaten Yes, current master works fine for me.

**Comment by NielsRogge** [comment]
@cgawron for me it also works fine with v4.15, not sure what the issue is?

**Pull request merged**